typing/typecore.ml

```diff
@@ -654,10 +650,10 @@ let solve_constructor_annotation env name_list sty ty_args ty_ex =
         {name with txt = id})
       name_list
   in
-  begin_def ();
-  let cty, ty, force = Typetexp.transl_simple_type_delayed !env sty in
-  end_def ();
-  generalize_structure ty;
+  let cty, ty, force =
+    with_local_level ~post:(fun (_,ty,_) -> generalize_structure ty)
+      (fun () -> Typetexp.transl_simple_type_delayed !env sty)
```

Comment:
I have a naive question about the usage of `Typetexp.transl_simple_type_delayed` in typecore.

Many of its occurrences are wrapped in a new level, with `generalize_structure` called on the `ty` comment (second element of the output triple). (The only two exceptions are the calls in `solve_Ppat_poly_constraint` and in the `Pexp_coerce` case.) But looking at the code of transl_simple_type_delayed, the middle component is the result of calling `instance` on `cty.ctyp_type`. What is the point of taking an instance and then re-generalizing right after ?

(Should this common behavior be wrapped in a new function of Typetexp, to simplify the logic in Typecore?)

Reply:
Good question. I don't have a full answer yet, but a first remark is that the generalization function used inside `transl_simple_type_delayed` is not the same as the one used outside. The internal one generalizes everything, but the external one doesn't generalize type variables, so it would be incorrect to just use the type inside `ctyp`. I assume that the reason to generalize inside `transl_simple_type_delayed` is to keep a generalized version of the type inside `ctyp`, but I have not yet found where it is used. The goal might just be to be able to print the annotation independently of types inferred from outside.